### PR TITLE
one hot encoder commited

### PR DESCRIPTION
**Pull request by NandiniGera**

### CODE REVIEW

**Review by harshjohar — CHANGES_REQUESTED** [review]
You dont need to import "vector" or "map", just include "core.hpp" file in this .hpp file. It has all the prerequisites.

**Comment by NandiniGera** [comment]
> You dont need to import "vector" or "map", just include "core.hpp" file in this .hpp file. It has all the prerequisites.

Done

File: src/slowmokit/methods/preprocessing/one_hot_encoder.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#include "one_hot_encoder.hpp"
+
+template<class T>
+
+// Function to one hot encode a vector of integers
+std::vector<std::vector<int>> one_hot_encoder(const std::vector<int> &data, int n_classes) {
```
Comment (uttammittal02) [inline src/slowmokit/methods/preprocessing/one_hot_encoder.cpp:6]: The data column might contain strings so better use template T. Remember we do one hot encoding for categorical variables

File: src/slowmokit/methods/preprocessing/one_hot_encoder.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  int class_index = 0;
+  // added each integer to class mapping
+  for (int value : data) {
+    if (class_mapping.count(value) == 0) {
+      class_mapping[value] = class_index++;
+    }
+  }
+
+  // Vector to store the one hot encoded data
+  std::vector<std::vector<int>> one_hot_encoded_data;
```
Comment (uttammittal02) [inline src/slowmokit/methods/preprocessing/one_hot_encoder.cpp:18]: Declare the vector size here itself, pushback might be slower

File: src/slowmokit/methods/preprocessing/one_hot_encoder.hpp [hunk 3]
```diff
@@ -0,0 +1,13 @@
+/**
+ * @file methods/preprocessing/one_hot_encoder.hpp
+ *
+ * Easy include To Return the one hot encoded data
+ */
+
+#ifndef ONE_HOT_ENCODER_HPP
+#define ONE_HOT_ENCODER_HPP
+#include "../../core.hpp"
+template<class T>
```
Comment (uttammittal02) [inline src/slowmokit/methods/preprocessing/one_hot_encoder.hpp:10]: Add documentation here i.e description of the parameters and its return value.
Screenshot 2023-02-06 at 10 24 09 PM

**Review by uttammittal02 — CHANGES_REQUESTED** [review]
Kindly make these changes in the concerned files. Also add an example [here](https://github.com/PEC-CSS/slowmokit/tree/main/examples/preprocessing)

**Review by ken1000minus7 — CHANGES_REQUESTED** [review]
Use camel cases for variables and functions

File: src/slowmokit/methods/preprocessing/one_hot_encoder.cpp (hunk 1 above)
Comment (NandiniGera) [reply src/slowmokit/methods/preprocessing/one_hot_encoder.cpp]: done

File: src/slowmokit/methods/preprocessing/one_hot_encoder.cpp (hunk 2 above)
Comment (NandiniGera) [reply src/slowmokit/methods/preprocessing/one_hot_encoder.cpp]: done

File: src/slowmokit/methods/preprocessing/one_hot_encoder.hpp (hunk 3 above)
Comment (NandiniGera) [reply src/slowmokit/methods/preprocessing/one_hot_encoder.hpp:19]: done

**Review by Ishwarendra — CHANGES_REQUESTED** [review]
Remove this and other comments (Line-4, 6, 14, 16, 20). They are redundant.

**Comment by NandiniGera** [comment]
> Remove this and other comments (Line-4, 6, 14, 16, 20). They are redundant.

done

**Comment by Ishwarendra** [comment]
@NandiniGera Few minor things to edit:
- Remove `@template T` from Line-13 in [one_hot_encoder.hpp](https://github.com/PEC-CSS/slowmokit/pull/71/files#diff-24fbfb682628f323648898934de1c0b801f6077d307c7dc4d2809280833fec44)
- In  [one_hot_encoder.hpp](https://github.com/PEC-CSS/slowmokit/pull/71/files#diff-24fbfb682628f323648898934de1c0b801f6077d307c7dc4d2809280833fec44) replace function signature at Line-20
    - `std::vector<std::vector<int>> one_hot_encoder(const std::vector<int> &data, int n_classes);` with
    -  `std::vector<std::vector<int>> one_hot_encoder(const std::vector<int> &, int );` (parameter names are not needed here)
- Add below comment in [one_hot_encoder.cpp](https://github.com/PEC-CSS/slowmokit/pull/71/files#diff-12e9b1ef64f011abd4583e379bdb0d6608f143734a077d68d95b2fb03399aefe) 

<details>
<summary>Add this comment at the start of one_hot_encoder.cpp</summary>

```cpp
/**
 * @file methods/preprocessing/one_hot_encoder.cpp
 *
 * Implementation of one hot encoder
 */
```
</details>

Other than this everything looks great to me! Nice job.

**Comment by NandiniGera** [comment]
> @NandiniGera Few minor things to edit:
> 
> * Remove `@template T` from Line-13 in [one_hot_encoder.hpp](https://github.com/PEC-CSS/slowmokit/pull/71/files#diff-24fbfb682628f323648898934de1c0b801f6077d307c7dc4d2809280833fec44)
> * In  [one_hot_encoder.hpp](https://github.com/PEC-CSS/slowmokit/pull/71/files#diff-24fbfb682628f323648898934de1c0b801f6077d307c7dc4d2809280833fec44) replace function signature at Line-20
>   
>   * `std::vector<std::vector<int>> one_hot_encoder(const std::vector<int> &data, int n_classes);` with
>   * `std::vector<std::vector<int>> one_hot_encoder(const std::vector<int> &, int );` (parameter names are not needed here)
> * Add below comment in [one_hot_encoder.cpp](https://github.com/PEC-CSS/slowmokit/pull/71/files#diff-12e9b1ef64f011abd4583e379bdb0d6608f143734a077d68d95b2fb03399aefe)
> 
> Add this comment at the start of one_hot_encoder.cpp
> Other than this everything looks great to me! Nice job.

done, thank youu :)

**Review by Ishwarendra — APPROVED** [review]

**Review by harshjohar — APPROVED** [review]
fine by me

**Review by uttammittal02 — APPROVED** [review]
Lgtm

File: CMakeLists.txt [hunk 4]
```diff
@@ -39,4 +39,6 @@ add_library(slowmokit
         src/slowmokit/methods/preprocessing/standardization.cpp
         src/slowmokit/methods/neighbors/bernoulli_nb.hpp
         src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.hpp
-        src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.cpp)
\ No newline at end of file
+        src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.cpp
```
Comment (uttammittal02) [inline CMakeLists.txt:42]: Once run the command `git pull upstream main` to fetch the new changes

**Review by uttammittal02 — CHANGES_REQUESTED** [review]

File: src/slowmokit/methods/preprocessing/one_hot_encoder.hpp [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * oneHotEncoder 
+ * @param data {const std::vector<T> &} - vector that contains the data that has to be one hot encoded
+ * @param nClasses {int} - integer that indicates the number of classes in the data.
+ * @return {std::vector<std::vector<int>>} - oneHotEncoded data is returned. 
+ */
+
+template<class T>
+std::vector<std::vector<int>> one_hot_encoder(const std::vector<int> &, int );
```
Comment (ken1000minus7) [inline src/slowmokit/methods/preprocessing/one_hot_encoder.hpp:19]: Fix the case of the function here

**Review by ken1000minus7 — CHANGES_REQUESTED** [review]

File: src/slowmokit/methods/preprocessing/one_hot_encoder.hpp (hunk 5 above)
Comment (NandiniGera) [reply src/slowmokit/methods/preprocessing/one_hot_encoder.hpp]: > Fix the case of the function here

done

**Comment by ken1000minus7** [comment]
Take a fresh pull from main branch as a pr was merged just now so there r conflicts

**Comment by NandiniGera** [comment]
> Take a fresh pull from main branch as a pr was merged just now so there r conflicts

done

File: CMakeLists.txt [hunk 6]
```diff
@@ -40,8 +40,5 @@ add_library(slowmokit
         src/slowmokit/methods/neighbors/bernoulli_nb.hpp
         src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.hpp
         src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.cpp
-        src/slowmokit/methods/metrics/classification_report.hpp
-        src/slowmokit/methods/metrics/classification_report.cpp
-        src/slowmokit/methods/neighbors/gaussian_nb.hpp
-        src/slowmokit/methods/neighbors/gaussian_nb/gaussian_nb.cpp
-        src/slowmokit/methods/neighbors/gaussian_nb/gaussian_nb.hpp)
```
Comment (ken1000minus7) [inline CMakeLists.txt:47]: I guess these lines got deleted when you pulled, add these above your files

Comment (NandiniGera) [reply CMakeLists.txt]: > I guess these lines got deleted when you pulled, add these above your files

done

**Review by ken1000minus7 — APPROVED** [review]
Lgtm, great work

File: src/slowmokit/methods/preprocessing/one_hot_encoder.hpp [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * oneHotEncoder 
+ * @param data {const std::vector<T> &} - vector that contains the data that has to be one hot encoded
+ * @param nClasses {int} - integer that indicates the number of classes in the data.
+ * @return {std::vector<std::vector<int>>} - oneHotEncoded data is returned. 
+ */
+
+template<class T>
+std::vector<std::vector<int>> oneHotEncoder(const std::vector<int> &, int );
```
Comment (uttammittal02) [inline src/slowmokit/methods/preprocessing/one_hot_encoder.hpp:19]: Here the vector should be passed as vector<T> and not vector<int>

**Review by uttammittal02 — CHANGES_REQUESTED** [review]

File: src/slowmokit/methods/preprocessing/one_hot_encoder.hpp (hunk 7 above)
Comment (NandiniGera) [reply src/slowmokit/methods/preprocessing/one_hot_encoder.hpp]: > Here the vector should be passed as `vector<T>` and not `vector<int>`

done

**Review by uttammittal02 — APPROVED** [review]
Lgtm

**Comment by harshjohar** [comment]
And finally it is approved by everyone 😂
Gg @NandiniGera

**Pull request merged**